### DISCUSSION TITLE
tutorials/animation/2d_skeletons

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#101] [User-contributed notes]
# tutorials/animation/2d_skeletons

Introduction: When working with 3D, skeletal deforms are common for characters and creatures and most 3D modeling applications support it. For 2D, as this function is not used as often, it's diffic...

https://docs.godotengine.org/en/stable/tutorials/animation/2d_skeletons.html

**Comment by thathurtabit** [comment]
Hmm, I'm up to the part:

```
Create a Bone2D node as a child of the skeleton. Put it on the hip (usually skeletons start here). The bone will be pointing to the right, but you can ignore this for now.
```

But when I add a `Bone2D` node (or several) nothing visually shows up - i.e. `the bone will be pointing to the right` - but nothing is visually appearing.

Maybe I'm missing something? I've tried Godot 4.2.2 and 4.3

**Reply by dodopod** [reply]
One possibility: in the toolbar above the viewport, there should be a bone icon with the tooltip "Skeleton Options" (it's 2 to the left of "View"). Try clicking on that, and then "Show Bones".

**Reply by thathurtabit** [reply]
Hi, thanks for the reply. I did try this and unfortunately, nothing seems to change/appear in the main viewport.

Also, in the `Inspector` tab, with a `Bone2D` node selected in the `Scene` tab, I've tried toggling `Inspector -> Editor Settings -> Show Bone Gizmo` but that doesn't seem to have an effect either.

https://github.com/user-attachments/assets/46b4c124-bec1-4613-b5de-de29074a3705

**Reply by dodopod** [reply]
If you create a new project, and add a Bone2D to an empty scene, does it show up then?

**Comment by thathurtabit** [comment]
I spotted the "issue", the issue is I'm new to Godot and an idiot.

After a while, I noticed that when clicking on the `Bone2D` node in the `Scene` view, the arrow colours changed next to where the `Root` nodes are. Essentially Godot is doing the right thing by placing new nodes at `0,0` position, and I didn't see it because that's the same position as the parent nodes 🤦 . 

Manually moving the `Bone2D` nodes with the `Move Mode (W)` lets me put them in the correct position.

Thanks for the suggestions!

<img width="495" alt="Screenshot 2024-08-31 at 16 52 25" src="https://github.com/user-attachments/assets/bacb8ea6-d85d-481d-8e11-6e02fb189d0e">

**Comment by AlinaEisert** [comment]
Hello there. I'm new to godot too and am struggeling with syncing the bones to the polygons, to be mor percise: the end bones. In my 2D Overview I see the 2D bones correctly rotated and everything. In the Inspector under "Bone2D" I unticked the "Auto Calculate Length and Angle" setting, rotated and defined the length manually so it would fit the part I'm animating. When I want to sync the bone to the polygon in the UV overview tho, I get a bone that is just set to it's standard settings (all endbones are pointing to the right, only the length is according to what I defined). Is this problematic if you want to animate without bending a polygon or in general or can I continue despite this happening? And how can I prevent that in future?

**Reply by AlinaEisert** [reply]
Furthermore: I discovered, that some parts are not moving the way they're supposed to, e.g. when I move the back arm at it's shoulder, the hand is moving according to the shoulder. However, the lower arm is moving in a kind of 3D way, which makes it become smaller the more up i move the arm by its shoulder. I already checked all the drawn weights, but the back arm has the same settings as the front arm: No further white dots are marked exceot for the according bodyparts (I did not quite understand the weight-stuff yet tho).

**Reply by redtrumpet** [reply]
I also had the problem with the endpoints. If you use a "Rotation" instead of the "Bone Angle" property, it shows up correctly.

**Reply by AlinaEisert** [reply]
thanks. I still don't get the weird transformation of my bones ... 
Can someone help? I have searched the internet far and wide and have found nothing regarding weird bone transforms when it comes to godot 2D animations ...

**Comment by rohans62** [comment]
Hi,
This is all well and good.  The question remains: now that we have our skele and mesh all working and have created great animations for them, HOW do you connect it to something that can be controlled by a characterbody2d so it is moveable (either ai or player) and will play the animations???

Everything I see shows how to tie sprite sheets to animation player but the whole point of rigging is to create better animations than sprite sheets.  I went through all the trouble of slicing, polygon-ing, skeletizing and creating animations.  Now the only way to use the animationplayer is to create sprite sheet poses? WTH?  Hopefully I am just not seeing the oblivious but I have looked in a lot of places and everything ties back to sprite sheet poses. 

All I am doing presently is generating one non-movable animation AND a movable (non-animated) sprite2d.

**Reply by Jeidoz** [reply]
Animation player supports multiple tracks per animation. You just need create a new animation, start recording changes and move bones to desired final position. For example, check this time stamp on [video](https://youtu.be/GQLL7laFqQY?si=tNyN0CcOEV5HWFJ-&t=562). Later on, you may want to use AnimationTree with Blend features to make a smooth transition between different animations and states.

**Comment by themipper** [comment]
I followed the tutorial and everything worked. But I have a different question regarding the usage of polygons, skeletons and animations.

Is there a way to flip the created animation horizontally?
Is there a way to scale the whole character?

I tried using the scale parameter but that only scales the skeleton and not the polygons. Which results in a skewed animation and character.

When using sprites I was using sprite.flip_h to flip it. Does something similar exist for the polygons and Skeleton?

**Reply by dodopod** [reply]
Try [SoupIK](https://godotengine.org/asset-library/asset/2472). I'm not sure how to do it without a plugin, though.

**Reply by themipper** [reply]
I found the solution. User error.

The polygons need to be nested under a Node2D. I used a Node which has no Transform.

Once I switched to Node2D as the root for the polygons everything works as expected. Scaling and flipping (X*-1).

**Reply by rohans62** [reply]
yep, that is the issue.

**Comment by rohans62** [comment]
update:  I did get it to work with this:

func _on_timer_timeout():
	timer_count += 1
	if timer_count % 2 == 0:
		player_node.position = Vector2(80,-80)
		player_node.scale.x = -1
	else:
		player_node.position = Vector2(-80,-80)
		player_node.scale.x = 1

forget about the timer stuff it was just for testing... but the concept being that I had to constantly set the position as well as the scale.x to keep it in one place.  is this the only way.  how efficient is it. I wonder.  It seems to be O(1) so that is pretty good. 

Hi folks,

for a 128x128 px characterbody2d setup like so:
characterbody2d
-collisionshape2d
-node2d
---polygon2d
----polygon2d (one for each texture - arms, legs, etc.)
---skeleton2d
----bone2d (one for each in the whole bone heirarchy thing)
-animationplayer

I am having a heck of a time trying to figure out how to flip_h or scale.x this.
I am able to with the node2d  and scale.x -1 
but it flips it from the top left corner so the characterbody2d doesn't actually stay in place and flip it moves 128px on the filp

you know how the sprite2d and animatedsprite2d flip in place easily and are used in movement stuff.  this seems not the case when using polygons and bones and animation player making me think I have to create completely new animations for each opposite movement direction.  This doesn't make sense so I must be (once again) missing something.

Any thoughts?

**Reply by rohans62** [reply]
okay so I got this working.  If I am missing something obviously better or built in on flipping skeletal and polygon rigged 2d chars lmk.

```gdscript
func move_body(direction: Vector2, _delta: float) -> void:
       if direction.x != _last_direction_x:
	  body.velocity = direction.normalized() * speed
	  body.move_and_slide()
	  sprite.position = Vector2(80,-80) if direction.x < 0 else Vector2(-80,-80)
	  sprite.scale.x = -1 if direction.x < 0 else 1
	  _last_direction_x = direction.x
```

**Comment by cofeek-codes** [comment]
I completed tutorial just fine, but how to add colliders that will move with bones now? Should `CollisionPolygon` be child of a bone?

**Reply by shadowfinderstudios** [reply]
Sorry for the late reply.

- CharacterBody2D
  - Skeleton2D
    - Bone2D ("Hand")
      - Area2D "Hitbox_Hand"
        - CollisionShape2D or CollisionPolygon2D
    - Bone2D "Torso"
      - Area2D "Hurtbox_Torso"
        - CollisionShape2D

For physics/ragdoll use PhysicalBone2D instead of collision shapes.

**Comment by Torisk-Ablinsky** [comment]
How the hell do I set up the tris to make my leg not look like a mangled mess :sob:

**Reply by Torisk-Ablinsky** [reply]
NVM I just realized it put the points on the wrong part of the uv lol :blush:

**Comment by AuthorInabsolutes** [comment]
Likely not everyone needs this much detail, but for those like my who didn't register the "toolbar" part of the tutorial's direction....
the "rest" position can be set on the Skeletal2D node (for it's children nodes) by switching to the 2D view. At the right side of the toolbar along the top of the view, it says Skeletal2D. Left click drops down a menu with just 2 options: "reset to rest position" and "Overwrite rest position"
